README.md
# GeeTestPOC

## Prerequisites

In order to test this POC locally you need to have a `.env` like its shown in [.env.example](./.env.example).

## Testing locally

There are two solvers GeeTest captcha in the present project, the slider solver and the space one. These POC are based on the following tutorials:

- [GeeTest slider solver](https://scraperbox.com/blog/solving-a-geetest-slider-captcha-with-puppeteer)
- [GeeTest space solver](https://www.youtube.com/watch?v=wPU8BTh5vKk)

In order to see there in action just run the following in your terminal:

```bash
pnpm run test:local
```

This will run a suit of unit test that will try to resolve the two different GeeTest captchas in different sites.

In case you want to test the slider or space captcha only, please run the following in your terminal

```bash
pnpm run test:X:local
```

Where "X" can be replaced with `space` or `slider`.

## Observations

Notice that currently the POC for the space captcha is not complete, since the page we are trying to solve its captcha is blocking us, we can go further to see it works or not. Because of this the `space` unit test will always be successful, even when the can not go further as mentioned before.


CHANGELOG.md
# GeeTestPOC

test/index.test.ts
import { solver as sliderSolver } from '../src/solver/slider'

describe('GeeTestPOC tests', () => {
  it('Must solve the GeeTest slider test', async () => {
    const result = await sliderSolver()

    expect(result).toBe(true)
  }, 45_000)
})


test/space.test.ts
import { solver } from '../src/solver/space'

describe('GeeTestPOC tests', () => {
  it('Must solve the GeeTest space test', async () => {
    const result = await solver()

    expect(result).toBe(true)
  }, 300_000)
})


test/slider.test.ts
import { solver } from '../src/solver/slider'

describe('GeeTestPOC tests', () => {
  it('Must solve the GeeTest slider test', async () => {
    const result = await solver()

    expect(result).toBe(true)
  }, 45_000)
})


src/solver/space.ts
import { Page } from 'puppeteer'
import axios from 'axios'
import { createIncognitoBrowser, navigateTo, sleep } from './utils'

const SITE_URL = 'https://nuwber.com/'
const LOAD_CAPTCHA_URL_CONTENT = 'get.php'
const LOAD_CAPTCHA_URL_CONTENT_RESPONSE = 'GEE'
const API_KEY = process.env.API_KEY

type GeeTestValues = {
  challenge: string
  gt: string
}

type SolvedCaptcha = {
  geetest_challenge: string
  geetest_validate: string
  geetest_seccode: string
}

const solveGeeTest = async (geeTestValues: GeeTestValues) => {
  const url = `http://2captcha.com/in.php?key=${API_KEY}&method=geetest&gt=${geeTestValues.gt}&challenge=${geeTestValues.challenge}&pageurl=${SITE_URL}&json=1`
  const { data } = await axios.get(url)
  console.log('data', data)

  const captchaId = data.request
  let complete = false
  let attempt = 0
  let captchaJson:
    | {
        status: number
        request: string | SolvedCaptcha
      }
    | undefined

  // It takes a bit of time to complete the geetest so we have to poll for it.
  // 175 seconds seems like enough (35 * 5)
  while (!complete && attempt < 35) {
    await sleep(5_000)
    attempt++
    const captchaUrl = `http://2captcha.com/res.php?key=${API_KEY}&action=get&id=${captchaId}&json=1`
    const captchaResponse = await axios.get(captchaUrl)
    captchaJson = captchaResponse.data

    if (
      captchaJson?.status === 1 ||
      captchaJson?.request !== 'CAPCHA_NOT_READY'
    )
      complete = true
  }

  return captchaJson?.request
}

const handleGeeTest = async (page: Page, geeTestValues: GeeTestValues) => {
  const iframeSrc = await page.$eval('iframe[id="main-iframe"]', element =>
    element.getAttribute('src')
  )
  const dai = iframeSrc?.match(/incident_id=([^&]*)/)?.[1]?.split('-')[1]
  const cts = iframeSrc?.match(/cts=([^&]*)/)?.[1]

  if (!dai || !cts) throw new Error('Failed to get dai or cts from iframe src')

  const solvedCaptcha = await solveGeeTest(geeTestValues)

  if (typeof solvedCaptcha === 'string')
    throw new Error(
      'Something went wrong, solvedCaptcha should not be a string'
    )

  if (!solvedCaptcha?.geetest_challenge)
    throw new Error('failed to solve captcha')

  await page.evaluate(
    (solvedCaptcha: SolvedCaptcha, dai: string, cts: string) => {
      if (!window.XMLHttpRequest)
        throw new Error('window.XMLHttpRequest is not defined')

      // This is their code that I copied so I could perfectly emulate how they submitted these values.
      const xhr = new window.XMLHttpRequest()
      const postBody =
        'geetest_challenge=' +
        solvedCaptcha.geetest_challenge +
        '&geetest_validate=' +
        solvedCaptcha.geetest_validate +
        '&geetest_seccode=' +
        solvedCaptcha.geetest_seccode

      xhr.open(
        'POST',
        `/_Incapsula_Resource?SWCGHOEL=gee&dai=${dai}&cts=${cts}`,
        true
      )
      xhr.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded')
      xhr.onreadystatechange = function (event) {
        console.log('ev', event)

        if (xhr.readyState === 4) window.parent.location.reload()
      }
      xhr.send(postBody)
    },
    solvedCaptcha,
    dai,
    cts
  )

  await sleep(1500)
  await page.goto(SITE_URL)
}

const solver = async () => {
  const { browser, page } = await createIncognitoBrowser()
  let firstTryJSBlock = true
  let firstTry = true
  let geeTestValues: GeeTestValues | undefined

  await page.setRequestInterception(true)

  page.on('request', async request => {
    if (request.url().includes(LOAD_CAPTCHA_URL_CONTENT) && firstTryJSBlock) {
      await request.abort()
      firstTryJSBlock = false
    } else await request.continue()
  })

  page.on('response', async response => {
    if (
      response.url().includes(LOAD_CAPTCHA_URL_CONTENT_RESPONSE) &&
      firstTry
    ) {
      console.log(
        '🚀 ~ file: space.ts:42 ~ solver ~ response.url:',
        response.url()
      )

      firstTry = false
      try {
        geeTestValues = await page.evaluate((url: string) => {
          return fetch(url)
            .then(response => response.json())
            .then(data => data)
        }, response.url())
      } catch (e) {
        console.log('🚀 ~ file: space.ts:47 ~ solver ~ e:', e)
      }
    }
  })

  await navigateTo({
    page,
    url: SITE_URL
  })

  const captchaIframe = await page.$('#main-iframe')

  if (captchaIframe && geeTestValues)
    try {
      await sleep(5_000)
      await handleGeeTest(page, geeTestValues)
    } catch (e) {
      await browser.close()

      if (!geeTestValues)
        console.log('geeTestValues is undefined, something went wrong')
    }

  await browser.close()

  return true
}

export { solver }


src/solver/utils.ts
import puppeteer, { Page } from 'puppeteer'

const SLIDE_CAPTCHA_URL = 'https://www.geetest.com/en/demo'
const GEE_TEST_SELECTOR = '.geetest_radar_tip'

const sleep = (time: number) =>
  new Promise(resolve => setTimeout(resolve, time))

const navigateTo = async ({
  page,
  url,
  selector
}: {
  page: Page
  url: string
  selector?: string
}) => {
  await page.goto(url, {
    timeout: 60_000,
    waitUntil: 'networkidle2'
  })

  if (selector) {
    await page.waitForSelector(selector)
    await page.click(selector)
    await sleep(5_000)
  }
}

const createIncognitoBrowser = async () => {
  const browser = await puppeteer.launch({
    headless: process.env.NODE_ENV !== 'local',
    args: [
      '--incognito',
      '--disable-web-security',
      '--disable-features=IsolateOrigins,site-per-process'
    ],
    devtools: true
  })
  const context = await browser.createIncognitoBrowserContext()
  const page = await context.newPage()

  return { browser, page }
}

export {
  SLIDE_CAPTCHA_URL,
  GEE_TEST_SELECTOR,
  createIncognitoBrowser,
  navigateTo,
  sleep
}


src/solver/slider.ts
import { Page } from 'puppeteer'
import Jimp from 'jimp'
import pixelmatch from 'pixelmatch'
import { cv } from 'opencv-wasm'

import {
  SLIDE_CAPTCHA_URL,
  createIncognitoBrowser,
  navigateTo,
  sleep
} from './utils'

const SLIDE_CAPTCHA_SELECTOR = '.tab-item.tab-item-1'
const GEE_TEST_SELECTOR = '.geetest_radar_tip'
const GEE_TEST_CANVAS = '.geetest_canvas_img canvas'
const SLIDER_SELECTOR = '.geetest_slider_button'
const GEE_TEST_RESULT =
  '.geetest_success_radar_tip > span.geetest_success_radar_tip_content'

const clickVerifyButton = async (page: Page) => {
  await page.click(GEE_TEST_SELECTOR)
  await page.waitForSelector(GEE_TEST_CANVAS, { visible: true })
  await sleep(1_000)
}

const verifyCaptchaResolved = async (page: Page) => {
  try {
    await page.waitForSelector(GEE_TEST_RESULT, {
      visible: true,
      timeout: 10_000
    })

    return true
  } catch (error) {
    return false
  }
}

const getCaptchaImages = async (page: Page) => {
  const images = await page.$$eval('.geetest_canvas_img canvas', canvases => {
    return canvases.map(canvas => {
      // This will get the base64 image data from the html canvas.
      // The replace function simply strip the "data:image" prefix.
      return canvas.toDataURL().replace(/^data:image\/png;base64,/, '')
    })
  })

  // For each base64 string create a Javascript buffer.
  const buffers = images.map(img => Buffer.from(img, 'base64'))

  // And read each buffer into a Jimp image.
  return {
    captcha: await Jimp.read(buffers[0]),
    puzzle: await Jimp.read(buffers[1]),
    original: await Jimp.read(buffers[2])
  }
}

type Images = Awaited<ReturnType<typeof getCaptchaImages>>

const getDiffImage = async (images: Images) => {
  const { width, height } = images.original.bitmap

  // Use the pixelmatch package to create an image diff
  const diffImage = new Jimp(width, height)

  pixelmatch(
    images.original.bitmap.data,
    images.captcha.bitmap.data,
    diffImage.bitmap.data,
    width,
    height,
    { includeAA: true, threshold: 0.2 }
  )

  // Use opencv to make the diff result more clear
  const src = cv.matFromImageData(diffImage.bitmap)
  const dst = new cv.Mat()
  const kernel = cv.Mat.ones(5, 5, cv.CV_8UC1)
  const anchor = new cv.Point(-1, -1)

  cv.threshold(src, dst, 127, 255, cv.THRESH_BINARY)
  cv.erode(dst, dst, kernel, anchor, 1)
  cv.dilate(dst, dst, kernel, anchor, 1)

  return new Jimp({
    width: dst.cols,
    height: dst.rows,
    data: Buffer.from(dst.data)
  })
}

type DiffImage = Awaited<ReturnType<typeof getDiffImage>>

const getPuzzlePieceSlotCenterPosition = (diffImage: DiffImage) => {
  const src = cv.matFromImageData(diffImage.bitmap)
  const dst = new cv.Mat()

  cv.cvtColor(src, src, cv.COLOR_BGR2GRAY)
  cv.threshold(src, dst, 150, 255, cv.THRESH_BINARY_INV)

  // This will find the contours of the image.
  const contours = new cv.MatVector()
  const hierarchy = new cv.Mat()

  cv.findContours(
    dst,
    contours,
    hierarchy,
    cv.RETR_EXTERNAL,
    cv.CHAIN_APPROX_SIMPLE
  )

  // Next, extract the center position from these contours.
  const contour = contours.get(0)
  const moment = cv.moments(contour)
  const cx = Math.floor(moment.m10 / moment.m00)
  const cy = Math.floor(moment.m01 / moment.m00)

  // Just for fun, let's draw the contours and center on a new image.
  cv.cvtColor(dst, dst, cv.COLOR_GRAY2BGR)

  const red = new cv.Scalar(255, 0, 0)

  cv.drawContours(dst, contours, 0, red)
  cv.circle(dst, new cv.Point(cx, cy), 3, red)

  return {
    x: cx,
    y: cy
  }
}

type Center = ReturnType<typeof getPuzzlePieceSlotCenterPosition>

const findMyPuzzlePiecePosition = async (page: Page) => {
  // Must call the getCaptchaImages again, because we have changed the
  // slider position (and therefore the image)
  const images = await getCaptchaImages(page)
  const srcPuzzleImage = images.puzzle
  const srcPuzzle = cv.matFromImageData(srcPuzzleImage.bitmap)
  const dstPuzzle = new cv.Mat()

  cv.cvtColor(srcPuzzle, srcPuzzle, cv.COLOR_BGR2GRAY)
  cv.threshold(srcPuzzle, dstPuzzle, 127, 255, cv.THRESH_BINARY)

  const kernel = cv.Mat.ones(5, 5, cv.CV_8UC1)
  const anchor = new cv.Point(-1, -1)

  cv.dilate(dstPuzzle, dstPuzzle, kernel, anchor, 1)
  cv.erode(dstPuzzle, dstPuzzle, kernel, anchor, 1)

  const contours = new cv.MatVector()
  const hierarchy = new cv.Mat()

  cv.findContours(
    dstPuzzle,
    contours,
    hierarchy,
    cv.RETR_EXTERNAL,
    cv.CHAIN_APPROX_SIMPLE
  )

  const contour = contours.get(0)
  const moment = cv.moments(contour)

  return {
    x: Math.floor(moment.m10 / moment.m00),
    y: Math.floor(moment.m01 / moment.m00)
  }
}

const slidePuzzlePiece = async (page: Page, center: Center) => {
  const sliderHandle = await page.$(SLIDER_SELECTOR)

  if (!sliderHandle) throw new Error('Could not find slider handle.')

  const handle = await sliderHandle.boundingBox()

  if (!handle) throw new Error('Could not find slider handle bounding box.')

  const handleX = handle.x + handle.width / 2
  const handleY = handle.y + handle.height / 2

  await page.mouse.move(handleX, handleY, { steps: 25 })
  await page.mouse.down()
  await sleep(250)

  let destX = handleX + center.x
  const destY = handleY + 32

  await page.mouse.move(destX, handleY, { steps: 25 })
  await sleep(100)

  // find the location of my puzzle piece.
  const puzzlePos = await findMyPuzzlePiecePosition(page)

  destX = destX + center.x - puzzlePos.x
  await page.mouse.move(destX, destY, { steps: 5 })
  await page.mouse.up()
}

const solver = async () => {
  const { browser, page } = await createIncognitoBrowser()

  await navigateTo({
    page,
    url: SLIDE_CAPTCHA_URL,
    selector: SLIDE_CAPTCHA_SELECTOR
  })
  await clickVerifyButton(page)

  if (await verifyCaptchaResolved(page)) return true

  const images = await getCaptchaImages(page)
  const diffImage = await getDiffImage(images)
  const center = getPuzzlePieceSlotCenterPosition(diffImage)

  await slidePuzzlePiece(page, center)
  await sleep(3_000)

  let result = false

  if (await verifyCaptchaResolved(page)) result = true

  await browser.close()

  return result
}

export { solver }


src/solver.ts
import { join } from 'node:path'
import { exec } from 'node:child_process'
import { promisify } from 'node:util'
import puppeteer, { Page } from 'puppeteer'
import Jimp from 'jimp'
import pixelmatch from 'pixelmatch'
import { cv } from 'opencv-wasm'

import { sleep } from 'utils'

const execAsync = promisify(exec)

const SLIDE_CAPTCHA_URL = 'https://www.geetest.com/en/demo'
const SLIDE_CAPTCHA_SELECTOR = '.tab-item.tab-item-1'
const GEETEST_SELECTOR = '.geetest_radar_tip'
const GEETEST_CANVAS = '.geetest_canvas_img canvas'
const SLIDER_SELECTOR = '.geetest_slider_button'

const navigateTo = async (page: Page) => {
  await page.goto(SLIDE_CAPTCHA_URL, {
    timeout: 60_000,
    waitUntil: 'networkidle2'
  })
  await page.waitForSelector(SLIDE_CAPTCHA_SELECTOR)
  await page.click(SLIDE_CAPTCHA_SELECTOR)
  await sleep(5_000)
}

const clickVerifyButton = async (page: Page) => {
  await page.click(GEETEST_SELECTOR)
  await page.waitForSelector(GEETEST_CANVAS, { visible: true })
  await sleep(1_000)
}

const getCaptchaImages = async (page: Page) => {
  const images = await page.$$eval('.geetest_canvas_img canvas', canvases => {
    return canvases.map(canvas => {
      // This will get the base64 image data from the html canvas.
      // The replace function simply strip the "data:image" prefix.
      return canvas.toDataURL().replace(/^data:image\/png;base64,/, '')
    })
  })

  // For each base64 string create a Javascript buffer.
  const buffers = images.map(img => Buffer.from(img, 'base64'))

  // And read each buffer into a Jimp image.
  return {
    captcha: await Jimp.read(buffers[0]),
    puzzle: await Jimp.read(buffers[1]),
    original: await Jimp.read(buffers[2])
  }
}

type Images = Awaited<ReturnType<typeof getCaptchaImages>>

const getDiffImage = async (images: Images) => {
  const { width, height } = images.original.bitmap

  // Use the pixelmatch package to create an image diff
  const diffImage = new Jimp(width, height)

  pixelmatch(
    images.original.bitmap.data,
    images.captcha.bitmap.data,
    diffImage.bitmap.data,
    width,
    height,
    { includeAA: true, threshold: 0.2 }
  )

  // Use opencv to make the diff result more clear
  const src = cv.matFromImageData(diffImage.bitmap)
  const dst = new cv.Mat()
  const kernel = cv.Mat.ones(5, 5, cv.CV_8UC1)
  const anchor = new cv.Point(-1, -1)

  cv.threshold(src, dst, 127, 255, cv.THRESH_BINARY)
  cv.erode(dst, dst, kernel, anchor, 1)
  cv.dilate(dst, dst, kernel, anchor, 1)

  return new Jimp({
    width: dst.cols,
    height: dst.rows,
    data: Buffer.from(dst.data)
  })
}

type DiffImage = Awaited<ReturnType<typeof getDiffImage>>

const getPuzzlePieceSlotCenterPosition = (diffImage: DiffImage) => {
  const src = cv.matFromImageData(diffImage.bitmap)
  const dst = new cv.Mat()

  cv.cvtColor(src, src, cv.COLOR_BGR2GRAY)
  cv.threshold(src, dst, 150, 255, cv.THRESH_BINARY_INV)

  // This will find the contours of the image.
  const contours = new cv.MatVector()
  const hierarchy = new cv.Mat()

  cv.findContours(
    dst,
    contours,
    hierarchy,
    cv.RETR_EXTERNAL,
    cv.CHAIN_APPROX_SIMPLE
  )

  // Next, extract the center position from these contours.
  const contour = contours.get(0)
  const moment = cv.moments(contour)
  const cx = Math.floor(moment.m10 / moment.m00)
  const cy = Math.floor(moment.m01 / moment.m00)

  // Just for fun, let's draw the contours and center on a new image.
  cv.cvtColor(dst, dst, cv.COLOR_GRAY2BGR)

  const red = new cv.Scalar(255, 0, 0)

  cv.drawContours(dst, contours, 0, red)
  cv.circle(dst, new cv.Point(cx, cy), 3, red)

  return {
    x: cx,
    y: cy
  }
}

type Center = ReturnType<typeof getPuzzlePieceSlotCenterPosition>

const findMyPuzzlePiecePosition = async (page: Page) => {
  // Must call the getCaptchaImages again, because we have changed the
  // slider position (and therefore the image)
  const images = await getCaptchaImages(page)
  const srcPuzzleImage = images.puzzle
  const srcPuzzle = cv.matFromImageData(srcPuzzleImage.bitmap)
  const dstPuzzle = new cv.Mat()

  cv.cvtColor(srcPuzzle, srcPuzzle, cv.COLOR_BGR2GRAY)
  cv.threshold(srcPuzzle, dstPuzzle, 127, 255, cv.THRESH_BINARY)

  const kernel = cv.Mat.ones(5, 5, cv.CV_8UC1)
  const anchor = new cv.Point(-1, -1)

  cv.dilate(dstPuzzle, dstPuzzle, kernel, anchor, 1)
  cv.erode(dstPuzzle, dstPuzzle, kernel, anchor, 1)

  const contours = new cv.MatVector()
  const hierarchy = new cv.Mat()

  cv.findContours(
    dstPuzzle,
    contours,
    hierarchy,
    cv.RETR_EXTERNAL,
    cv.CHAIN_APPROX_SIMPLE
  )

  const contour = contours.get(0)
  const moment = cv.moments(contour)

  return {
    x: Math.floor(moment.m10 / moment.m00),
    y: Math.floor(moment.m01 / moment.m00)
  }
}

const slidePuzzlePiece = async (page: Page, center: Center) => {
  const sliderHandle = await page.$(SLIDER_SELECTOR)

  if (!sliderHandle) throw new Error('Could not find slider handle.')

  const handle = await sliderHandle.boundingBox()

  if (!handle) throw new Error('Could not find slider handle bounding box.')

  const handleX = handle.x + handle.width / 2
  const handleY = handle.y + handle.height / 2

  await page.mouse.move(handleX, handleY, { steps: 25 })
  await page.mouse.down()
  await sleep(250)

  let destX = handleX + center.x
  const destY = handleY + 32

  await page.mouse.move(destX, handleY, { steps: 25 })
  await sleep(100)

  // find the location of my puzzle piece.
  const puzzlePos = await findMyPuzzlePiecePosition(page)

  destX = destX + center.x - puzzlePos.x
  await page.mouse.move(destX, destY, { steps: 5 })
  await page.mouse.up()
}

const removeImages = async () => {
  const path = join(__dirname, './images')

  await execAsync(`rm  ${path}/*`)
}

const solver = async () => {
  await removeImages().catch(() => {
    console.log('Nothing to delete')
  })
  const browser = await puppeteer.launch({
    headless: process.env.NODE_ENV !== 'local',
    args: ['--incognito']
  })
  const context = await browser.createIncognitoBrowserContext()
  const page = await context.newPage()

  await navigateTo(page)
  await clickVerifyButton(page)

  const images = await getCaptchaImages(page)
  const diffImage = await getDiffImage(images)
  const center = getPuzzlePieceSlotCenterPosition(diffImage)

  await slidePuzzlePiece(page, center)
  await sleep(3_000)
  await page.screenshot({
    path: join(__dirname, './images/solved.png'),
    fullPage: true
  })
  await browser.close()
}

export { solver }
